kvbc/test/v4blockchain/v4_keys_history_test.cpp

```diff
@@ -0,0 +1,446 @@
+
+  db->rawDB().CompactRange(::rocksdb::CompactRangeOptions{},
+                           db->columnFamilyHandle(v4blockchain::detail::KEYS_HISTORY_CF),
+                           &start_key,
+                           &end_key);
+
+  // make sure that keys of old blocks were filtered out from keys history
+  for (BlockId i = 1; i <= num_blocks; i++) {
+    for (int j = 0; j < num_keys; j++) {
+      auto key = std::to_string(j);
+      auto opt_val = keys_history.getVersionFromHistory(merkle_category, key, i);
+      if (i > num_blocks - num_blocks_to_save_in_history) {
+        ASSERT_TRUE(opt_val);
+        ASSERT_EQ(opt_val.value(), i);
+      } else {
+        ASSERT_FALSE(opt_val);
+      }
+    }
+  }
+
+  // check the case when keysHistoryMaxBlocksNum == 0
+  bftEngine::ReplicaConfig::instance().setkeysHistoryMaxBlocksNum(0);
+
+  for (BlockId i = num_blocks + 1; i <= 2 * num_blocks; i++) {
+    add_merkle_keys_to_block(keys_history, i, num_keys);
+  }
+  v4blockchain::detail::Blockchain::global_genesis_block_id = num_blocks + (BlockId)dist2(gen);
+
+  // compaction triggers filter
+  first_key = prefix + std::to_string(1) + v4blockchain::detail::Blockchain::generateKey(1);
+  last_key = prefix + std::to_string(2 * num_blocks) + v4blockchain::detail::Blockchain::generateKey(2 * num_blocks);
+  start_key = ::rocksdb::Slice(concord::storage::rocksdb::detail::toSlice(first_key));
+  end_key = ::rocksdb::Slice(concord::storage::rocksdb::detail::toSlice(last_key));
+
+  db->rawDB().CompactRange(::rocksdb::CompactRangeOptions{},
+                           db->columnFamilyHandle(v4blockchain::detail::KEYS_HISTORY_CF),
+                           &start_key,
+                           &end_key);
+
+  // make sure that keys of blocks below the genesis were filtered out from keys history
+  for (BlockId i = num_blocks + 1; i <= 2 * num_blocks; i++) {
+    for (int j = 0; j < num_keys; j++) {
+      auto key = std::to_string(j);
+      auto opt_val = keys_history.getVersionFromHistory(merkle_category, key, i);
+      if (i >= v4blockchain::detail::Blockchain::global_genesis_block_id) {
+        ASSERT_TRUE(opt_val);
+        ASSERT_EQ(opt_val.value(), i);
+      } else {
+        ASSERT_FALSE(opt_val);
+      }
+    }
+  }
+
+  // cleanup test changes
+  bftEngine::ReplicaConfig::instance().setkeysHistoryMaxBlocksNum(orig_max_blocks_to_save);
+  v4blockchain::detail::Blockchain::global_latest_block_id = 0;
+  v4blockchain::detail::Blockchain::global_genesis_block_id = 0;
+}
+
+TEST_F(v4_kvbc, follow_key_history) {
+  v4blockchain::detail::KeysHistory keys_history{db, categories};
+  std::random_device seed;
+  std::mt19937 gen{seed()};                       // seed the generator
+  std::uniform_int_distribution dist{10, 40};     // set min and max
+  const BlockId num_blocks = (BlockId)dist(gen);  // generate number
+  const int num_keys = dist(gen);
+  std::optional<BlockId> opt_val;
+  const auto key1 = "key1";
+
```

Comment:
And also query for key1 when the CF is empty and when it's populated with some keys but not with key1

Reply:
Added query when CF is empty. already checked when CF contains keys but not key1 (the first for loop)